File: app/models/facturas.py
from sqlalchemy import Column, Integer, BigInteger, String, Numeric, Text, DateTime, func
from sqlalchemy.orm import relationship

from app.database import Base

class Factura(Base):
    __tablename__ = "facturas"

    id = Column(Integer, primary_key=True, autoincrement=True)
    fecha = Column(DateTime, nullable=False, server_default=func.now())
    tercero_id = Column(Integer, nullable=False)
    vendedor_id = Column(Integer, nullable=False)
    resolucion_id = Column(Integer, nullable=False)
    prefijo = Column(String(10), nullable=False)
    consecutivo = Column(BigInteger, nullable=False)
    numero_completo = Column(String(30), nullable=False)
    forma_pago_id = Column(Integer, nullable=False)
    medio_pago_id = Column(Integer, nullable=False)
    subtotal = Column(Numeric(14,2), nullable=False, default=0)
    descuento_total = Column(Numeric(14,2), nullable=False, default=0)
    iva_total = Column(Numeric(14,2), nullable=False, default=0)
    total = Column(Numeric(14,2), nullable=False, default=0)
    notas = Column(Text, nullable=True)    
    usuario_creacion = Column(String(100), nullable=True)
    fecha_creacion = Column(DateTime, server_default=func.now())
    usuario_modifico = Column(String(100), nullable=True)
    fecha_modificacion = Column(DateTime, nullable=True)
    id_sucursal = Column(Integer, nullable=False)
    id_usuario = Column(Integer, nullable=False)
    
    detalles = relationship("FacturaDetalle", back_populates="factura", cascade="all, delete-orphan")

File: lib/api/facturas.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL;
import { authHeaders } from "@/lib/utils";

/* ============================================
   INTERFACES
============================================ */
export interface FacturaDetalleData {
  producto_id: number | null;
  presentacion_id?: number | null;
  variante_id?: number | null;
  descripcion: string;

  cantidad: number;
  precio_unitario: number;
  descuento: number;
  iva: number; // porcentaje enviado al backend
}

export interface FacturaData {
  tercero_id: number;
  vendedor_id: number;
  resolucion_id: number;

  prefijo: string;
  consecutivo: number;

  forma_pago_id: number;
  medio_pago_id: number;

  notas?: string;

  detalles: FacturaDetalleData[];
}

export interface Factura extends FacturaData {
  id: number;
  numero_completo: string;
  subtotal: number;
  descuento_total: number;
  iva_total: number;
  total: number;
  fecha_creacion: string;
  usuario_creacion?: string;
}

/* ============================================
   OBTENER TODAS LAS FACTURAS
============================================ */
export const obtenerFacturas = async (): Promise<Factura[] | null> => {
  try {
    const res = await fetch(`${API_URL}/facturas` ,{
      headers: authHeaders(),
    });

    if (!res.ok) {
      throw new Error(`Error al cargar facturas: ${res.status}`);
    }

    return await res.json();

  } catch (error) {
    console.error("Error obteniendo facturas:", error);
    return null;
  }
};

/* ============================================
   OBTENER UNA FACTURA POR ID
============================================ */
export const obtenerFactura = async (id: number): Promise<Factura | null> => {
  try {
    const res = await fetch(`${API_URL}/facturas/${id}` ,{
      headers: authHeaders(),
    });

    if (!res.ok) {
      throw new Error(`Factura no encontrada: ${res.status}`);
    }

    return await res.json();

  } catch (error) {
    console.error("Error obteniendo factura:", error);
    return null;
  }
};

/* ============================================
   CREAR FACTURA
============================================ */
export async function crearFactura(data: FacturaData) {
  try {
    
    console.log("➡️ DATA QUE LLEGA A crearFactura:", data)
    
    const response = await fetch(`${API_URL}/facturas`, {
      method: "POST",
      headers: authHeaders(),
      credentials: "include",
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => null);
      throw new Error(errorData?.detail || "Error al crear factura");
    }

    return await response.json();

  } catch (err) {
    console.error("Error en crearFactura:", err);
    throw err;
  }
}

/* ============================================
   ACTUALIZAR FACTURA
============================================ */
export async function actualizarFactura(id: number, data: FacturaData) {
  try {
    const response = await fetch(`${API_URL}/facturas/${id}`, {
      method: "PUT",
      headers: authHeaders(),
      credentials: "include",
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => null);
      throw new Error(errorData?.detail || "Error al actualizar factura");
    }

    return await response.json();

  } catch (err) {
    console.error("Error en actualizarFactura:", err);
    throw err;
  }
}

/* ============================================
   ELIMINAR FACTURA
============================================ */
export async function eliminarFactura(id: number) {
  try {
    const response = await fetch(`${API_URL}/facturas/${id}`, {
      method: "DELETE",
      credentials: "include",
      headers: authHeaders(),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => null);
      throw new Error(errorData?.detail || "Error al eliminar factura");
    }

    return true;

  } catch (err) {
    console.error("Error en eliminarFactura:", err);
    throw err;
  }
}
